Reapply "[Driver][ROCm][OpenMP] Fix default ockl linking for OpenMP."…

ampandey-1995 opened the pull request:
- This reverts commit https://github.com/llvm/llvm-project/commit/0c6c4a99936d4d39015c8d2332483f8db78f69cf.
  - Add '-mcode-object-version=5' as to explicitly use code object
    version 5 to match with 'FAIL' diagnostic.
  - Add Requires directive to support lit test run on platforms
    registered with x86_64 and amdgpu.

ampandey-1995 committed:
Reapply "[Driver][ROCm][OpenMP] Fix default ockl linking for OpenMP."
(#126628)

  - This reverts commit 0c6c4a99936d4d39015c8d2332483f8db78f69cf.
  - Add '-mcode-object-version=5' as to explicitly use code object
    version 5 to match with 'FAIL' diagnostic.
  - Add Requires directive to support lit test run on platforms
    registered with x86_64 and amdgpu.

arsenm commented on clang/lib/Driver/ToolChains/AMDGPU.cpp:
```diff
@@ -1014,17 +1014,20 @@ RocmInstallationDetector::getCommonBitcodeLibs(
-  auto GPUSanEnabled = [GPUSan]() { return std::get<bool>(GPUSan); };
+  // GPU Sanitizer currently only supports ASan and is enabled through host
+  // ASan.
+  auto GPUSanEnabled = [GPUSan]() {
+    return std::get<bool>(GPUSan) &&
+           std::get<const SanitizerArgs>(GPUSan).needsAsanRt();
+  };
```
Why is this using a lambda and capture for a simple and? I'm also surprised `get<bool>` works instead of `get<0>`.

ampandey-1995 replied:
> Why is this using a lambda and capture for a simple and?
If we have other sanitizers apart from asan such as msan for GPU. Then , what I think we would be having another msanrtl.bc linked so that future code would be something like below for easy readability and maintainability purpose. Note: we cannot enable both ASan and MSan simultaneously at compile time.  

```
auto GPUSanEnabled = [GPUSan](){
auto SanArgs = std::get<const SanitizerArgs>(GPUSan);
return std::get<bool>(GPUSan) &&
           (SanArgs.needsAsanRt() || SanArgs.needsMsanRt()) ;
};
```

> I'm also surprised `get<bool>` works instead of `get<0>`
Yes, since the  tuple contents are of unique types it works. Should I change it?

arsenm replied:
Just remove the lambda. It's over complicating a hypothetical future

jhuber6 replied:
The fact that we have `-fgpu-sanitize` at all is a mistake. The problem with these types of 'temporary' flags is that they come in and then never go away for legacy reasons. The correct solution would've been to force `-Xarch_host` or `-Xarch_device` if users didn't want it on both.

ampandey-1995 committed:
Address Comments of @arsenm.

jhuber6 commented on clang/lib/Driver/ToolChains/AMDGPU.cpp:
```diff
@@ -1014,25 +1014,27 @@ RocmInstallationDetector::getCommonBitcodeLibs(
-      auto SanArgs = std::get<const SanitizerArgs>(GPUSan);
-      if (SanArgs.needsAsanRt())
-        AddBCLib(getAsanRTLPath(), false);
+    if (GPUSanEnabled) {
+      AddBCLib(getAsanRTLPath(), false);
     }
```
```suggestion
    if (GPUSanEnabled)
      AddBCLib(getAsanRTLPath(), false);
```

ampandey-1995 replied:
Done.

ampandey-1995 committed:
Address @jhuber6 comments.

jhuber6 commented on clang/lib/Driver/ToolChains/AMDGPUOpenMP.cpp:
```diff
@@ -68,7 +68,9 @@ llvm::opt::DerivedArgList *AMDGPUOpenMPToolChain::TranslateArgs(
   DerivedArgList *DAL =
       HostTC.TranslateArgs(Args, BoundArch, DeviceOffloadKind);
-  if (!DAL || Args.hasArg(options::OPT_fsanitize_EQ))
+  // Skip sanitize options passed from the HostTC. The decision to instrument
+  // device code is computed only by 'shouldSkipSanitizeOption'.
+  if (!DAL && DAL->hasArg(options::OPT_fsanitize_EQ))
```
I don't fully understand why we need this. Shouldn't the `shouldSkip` part handle this just fine?

ampandey-1995 replied:
Sorry, actually here the DAL computed from the HostTC picks up the sanitize option like ```-fsanitize=address``` and pass forward to ```DAL``` of device toolchain. The ```shouldSkipSanitizeOption ```  runs over each argument of ```Args [The user provided arguments which are not derived]``` and checks if the ```-fsanitize=address``` is there it early returns false which then avoids checking the :xnack+ capability. 

So,  assume the  case where when we pass ```-fsanitize=address --offload-arch=gfx90a``` . This should avoid appending  ```-fsanitize=address``` in the DAL for device toolchain but it dosen't since the DAL already has -fsanitize=address from host toolchain and due to early return from ```shouldSkipSanitizeOption``` where it checks for the ```:xnack+``` capability is skipped.

ampandey-1995 replied:
What I have seen for HIPAMDToolChain , the DAL is always constructed every time dynamically allocated every time.

ampandey-1995 replied:
Should The ```shouldSkipSanitizeOption``` consider an extra parameter ```DAL``` allocated by the host and resolve that issue ? @jhuber6  and @yxsamliu  is that the correct way ,anything I am missing @yxsamliu ?

ampandey-1995 replied:
> Sorry I'm still confused. If the user cannot enable `-fsanitize=address` if the GPU target doesn't support `xnack` then that should be separate driver check / error. Unless we're just silently ignoring that?

Yes we are silently ignoring that and that ```-fsanitize=address``` is there for the host only, but for device it is skipped.

jhuber6 replied:
Silently ignoring this sounds pretty awful to me, but if that's what ROCm wants I guess that's the behavior.

jhuber6 replied:
If the goal is to ignore it, why can't we just instead be like `if (TCArgs.hasFlag(...) && isValid)` on every use instead of hacking around it in the argument list.

ampandey-1995 replied:
We actually in check it in this part of code https://github.com/llvm/llvm-project/blob/752ebec054f8b0bf7c221813724984f10096355a/clang/lib/Driver/ToolChains/AMDGPU.cpp#L1118

jhuber6 replied:
I just don't understand the rationale for removing it from the argument list instead of just guarding wherever we look up this flag.

ampandey-1995 replied:
> if it's not legal for some case, then emit an error. If the user doesn't want it then they can do `-Xarch-host -fsanitize=address`.

We actually emit the warning not error, since it's legal for the host to be asan instrumented but for device it depends based on ```--offload-arch=``` containing :xnack+. If :xnack+ then allow it otherwise emit warning silently skip it.

jhuber6 replied:
So, is it prohibitively difficult to do something more akin to.

```c
if (Args.hasFlag(OPT_fsanitize_address, OPT_fno_sanitize, false) &&
    Args.hasFlag(OPT_fgpu_sanitize, OPT_fno_gpu_sanitize, true)) {
  if (isSanitizerLegal(Args))
    // something
  else
    Diags.warn("incompatible, ignoring");    
}
```

jhuber6 reviewed (approved):
Good enough for now, thanks. Maybe some day we could stash the GPU asan code in `compiler-rt` somewhere so we can have upstream tests for this stuff.

ampandey-1995 committed:
[NFC] Remove std::tuple for GPUSan.

Don't require std::tuple<bool,const SanitizerArgs> parameter. Simplify
it to single boolean 'GPUSan'.

ampandey-1995 commented:
Hi @arsenm & @jhuber6  , I removed the usage of ```std::tuple<bool,const SanitizerArgs>``` in the last commit 
f738420 so as to simplify it much further.
